test/integration/job/job_test.go

```diff
@@ -1669,79 +1669,80 @@ func TestIndexedJob(t *testing.T) {
 }
 
 func TestJobPodReplacementPolicy(t *testing.T) {
-	const podCount int32 = 2
 	indexedCompletion := batchv1.IndexedCompletion
 	nonIndexedCompletion := batchv1.NonIndexedCompletion
 	var podReplacementPolicy = func(obj batchv1.PodReplacementPolicy) *batchv1.PodReplacementPolicy {
 		return &obj
 	}
-	jobSpecIndexedDefault := &batchv1.JobSpec{
-		Parallelism:    pointer.Int32Ptr(podCount),
-		Completions:    pointer.Int32Ptr(podCount),
-		CompletionMode: &indexedCompletion,
+	type jobStatus struct {
+		active      int
+		failed      int
+		terminating *int32
 	}
 	cases := map[string]struct {
 		podReplacementPolicyEnabled bool
-		deletePods                  bool
-		failPods                    bool
-		wantTerminating             *int32
-		wantFailed                  int
-		wantActive                  int
 		jobSpec                     *batchv1.JobSpec
+		wantStatusAfterDeletion     jobStatus
+		wantStatusAfterFailure      jobStatus
 	}{
-		"feature flag off, delete pods and verify no terminating status": {
-			deletePods: true,
-			jobSpec:    jobSpecIndexedDefault,
-			wantActive: int(podCount),
-			wantFailed: int(podCount),
-		},
-		"feature flag true, delete pods and verify terminating status": {
-			podReplacementPolicyEnabled: true,
-			deletePods:                  true,
-			jobSpec:                     jobSpecIndexedDefault,
-			wantTerminating:             pointer.Int32(podCount),
-			wantFailed:                  int(podCount),
+		"feature flag off, delete & fail pods, recreate terminating pods, and verify job status counters": {
+			jobSpec: &batchv1.JobSpec{
+				Parallelism:    ptr.To[int32](2),
+				Completions:    ptr.To[int32](2),
+				CompletionMode: &indexedCompletion,
+			},
+			wantStatusAfterDeletion: jobStatus{
+				active: 2,
+				failed: 2,
+			},
+			wantStatusAfterFailure: jobStatus{
+				active: 2,
+				failed: 2,
+			},
 		},
-		"feature flag true, delete pods, verify terminating status and recreate upon terminating": {
+		"feature flag true, TerminatingOrFailed policy, delete & fail pods, recreate terminating pods, and verify job status counters": {
 			podReplacementPolicyEnabled: true,
-			deletePods:                  true,
 			jobSpec: &batchv1.JobSpec{
-				Parallelism:          pointer.Int32Ptr(podCount),
-				Completions:          pointer.Int32Ptr(podCount),
+				Parallelism:          ptr.To[int32](2),
+				Completions:          ptr.To[int32](2),
 				CompletionMode:       &indexedCompletion,
 				PodReplacementPolicy: podReplacementPolicy(batchv1.TerminatingOrFailed),
 			},
-			wantTerminating: pointer.Int32(podCount),
-			wantFailed:      int(podCount),
+			wantStatusAfterDeletion: jobStatus{
+				active:      2,
+				failed:      2,
+				terminating: ptr.To[int32](2),
+			},
+			wantStatusAfterFailure: jobStatus{
+				active:      2,
+				failed:      2,
+				terminating: ptr.To[int32](2),
```

Comment:
did you investigate why this stays at 2?

Reply:
To some degree, this line https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L672 returns different results between the TerminatingOrFailed and Failed policy test cases, but I still cannot fully understand what is going on

Reply:
But even if it does, they should be filtered out here, no? https://github.com/kubernetes/kubernetes/blob/baea2df9d6f02935849e00390e0e8a5fad596f51/pkg/controller/job/job_controller.go#L790C46-L790C46

Reply:
here is the bug https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L925-L933

we set `job.Status.Terminating = jobCtx.terminating` on line 929 and then do the check `needsStatusUpdate = needsStatusUpdate || !ptr.Equal(job.Status.Terminating, jobCtx.terminating)` but at this point `!ptr.Equal(job.Status.Terminating, jobCtx.terminating)` will always return `false`.

If we just do a simple reorder, all tests work correctly and `terminating` count correctly drops to 0 for `TerminatingOrFailed` policy.
```
needsStatusUpdate := suspendCondChanged || active != job.Status.Active || !ptr.Equal(ready, job.Status.Ready)
needsStatusUpdate = needsStatusUpdate || !ptr.Equal(job.Status.Terminating, jobCtx.terminating)
job.Status.Active = active
job.Status.Ready = ready
job.Status.Terminating = jobCtx.terminating
err = jm.trackJobStatusAndRemoveFinalizers(ctx, jobCtx, needsStatusUpdate)
if err != nil {
	return fmt.Errorf("tracking status: %w", err)
}
```

/cc @mimowo @kannon92

Reply:
Is the only reason why this worked for Failed is because we had other updates around the same time?

In case of TerminatingOrFailed, we had a case where terminating was the only field changing.  While Failed had the failed or active field set?

This change looks good to me.  I think you can update it in this PR if you want.  We don't cherry-pick alpha features so I don't think we have to backport this.

Reply:
I guess it got decremented when the Job status got updated for some other reason, and the side-effect was the `terminating` count dropping to 0, but it was missing the update because of the `terminating` field being 0 earlier